### PR TITLE
[EGNN(C)] Convolution based EGNN layer

### PR DESCRIPTION
**Pull request by abigail-katcoff-kensho**
## Description
This is a Pytorch implementation of the graph neural network from [CVPR2019: Exploiting Edge Features for Graph Neural Network.](https://openaccess.thecvf.com/content_CVPR_2019/html/Gong_Exploiting_Edge_Features_for_Graph_Neural_Networks_CVPR_2019_paper.html)

There is an open issue requesting this implementation: https://github.com/dmlc/dgl/issues/2365

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/nn/pytorch/conv/edgegraphconv.py [hunk 1]
```diff
@@ -294,20 +248,15 @@ def forward(self, graph, feat, weight=None, edge_weight=None):
...
-                    feat_src = th.matmul(feat_src, weight)
-                graph.srcdata['h'] = feat_src
-                graph.update_all(aggregate_fn, fn.sum(msg='m', out='h'))
-                rst = graph.dstdata['h']
-            else:
-                # aggregate first then mult W
-                graph.srcdata['h'] = feat_src
-                graph.update_all(aggregate_fn, fn.sum(msg='m', out='h'))
-                rst = graph.dstdata['h']
+            single_channel_rsts = []
```
Comment (delphinevendryes-kensho) [inline python/dgl/nn/pytorch/conv/edgegraphconv.py]: because of the way EGNN works, I think we might require that graph is fully connected and the edge feature matrix is (N (N + 1) / 2, P) or (N (N - 1) / 2, P) depending on whether we have self edges. In that case, I think we can do a simple matmul operation rather than a for loop. What are your thoughts?

Comment (abigail-katcoff-kensho) [reply python/dgl/nn/pytorch/conv/edgegraphconv.py]: For a large graph, requiring that it be fully connected might be too strict -- why not let the user decide when it is appropriate to use this layer type? 

Pros of this current implementation (for loop):
more efficient 
matches the representation currently used for edge features 
Pros for matrix mul:
simpler to understand (for non-users of DGL)
avoids for loop

My ideal scenario would be to avoid this for loop by finding the correct DGL built-in to use.

File: python/dgl/nn/pytorch/conv/edgegraphconv.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+                else:
+                    norm = 1.0 / degs
+                shp = norm.shape + (1,) * (feat_dst.dim() - 1)
+                norm = th.reshape(norm, shp)
+                rst = rst * norm
+
+            if self._activation is not None:
+                rst = self._activation(rst)
+
+            return rst
```
Comment (delphinevendryes-kensho) [inline python/dgl/nn/pytorch/conv/edgegraphconv.py:276]: do we want to return the edge features here as well so they can be fed into the next layer?

Comment (abigail-katcoff-kensho) [reply python/dgl/nn/pytorch/conv/edgegraphconv.py:276]: The edge features don't change in edge convolution

File: python/dgl/nn/pytorch/conv/edgegraphconv.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+                                   'to be `True` when constructing this module will '
+                                   'suppress the check and let the code run.')
+            if edge_weight is None:
+                raise DGLError("Edge weight is required for this layer type (EdgeGraphConv)")
+            assert edge_weight.shape[0] == graph.number_of_edges()
+            aggregate_fn = fn.u_mul_e('h', '_edge_weight', 'm')
+
+            # (BarclayII) For RGCN on heterogeneous graphs we need to support GCN on bipartite.
+            feat_src, feat_dst = expand_as_pair(feat, graph)
+            if self._norm in ['left', 'both']:
```
Comment (delphinevendryes-kensho) [inline python/dgl/nn/pytorch/conv/edgegraphconv.py:233]: @abigail-katcoff-kensho  do you understand why they normalize on the left before the following operations and on the right after?
